### PR TITLE
Merge of Mbed Crypto

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
This is a merge of [Mbed Crypto](https://github.com/ARMmbed/mbed-crypto) into Mbed TLS. See ARMmbed/mbed-crypto#374.

Once this pull request is merged, we will merge any subsequent changes made in Mbed Crypto into Mbed TLS, then we will stop accepting pull requests to Mbed Crypto.

### Strategy

* Take `mbedtls/development`.
* Take `mbed-crypto/development`. Revert some commits that remove X509/TLS functionality; I call this “unremoval”. Make a few additional changes to facilitate the subsequent merge. Please use https://github.com/ARMmbed/mbed-crypto/pull/384 to review this part.
* Merge the crypto branch with the unremoval into the tls branch. For the merge commit, force all tls-specific files to their tls version even if the unremoval restored them to an older state. Force a few other files to their tls version where the crypto version is irrelevant. Resolve merge conflicts. See the merge commit message for details.
    The merge commit makes unwanted changes because some things have been removed in crypto due to being tls-only, and have not been touched in tls. Subsequent commits look for such things to restore.
* Restore the test coverage (`all.sh` and other scripts).
* Restore X509/TLS features (mainly a matter of fixing `config.h`).
* Undo some other accidental regressions due to the merge found by reviewing the diff.
* Remove the crypto submodule.
* Fix test failures:
    * :heavy_check_mark: `check-names.sh` fails because mbed-crypto rejects public identifiers containing `__` (double-underscore) and mbedtls defines such identifiers. Fix https://github.com/ARMmbed/mbedtls/issues/3096
    * :heavy_check_mark: Fix some build systems that got broken along the line.
    * :heavy_check_mark: Some test data from one side is missing dependencies that were only tested on the other side.
    * :heavy_check_mark: Ensure that there is no API change in the default configuration.
    * :information_source: **Mbed OS tests fail on the CI due to preexisting issues so this PR will not be tested with Mbed OS**.

https://github.com/ARMmbed/mbedtls/pull/3078 tries to reach the same point with a different approach: commit a minimum set of modified files in the merge commit, then selectively commit the remaining changes, analyzing their suitability. This has the benefit that unwanted changes are never made, whereas the approach here makes some unwanted changes in the merge commit then undoes them. But it makes the history very hard to follow.

### Correctness claim

**Claim: this merge preserves the features and test coverage of both Mbed Crypto and Mbed TLS.**

To verify this claim, divide the history in two parts:

1. The section comprising the merge commit and the restoration of things unduly removed or modified by the merge commit. This section ends at the commit Add changelog entries about the crypto submodule"".
2. The rest is comprised of fixes to bugs related to the merging of the two projects.

To verify that nothing from Mbed TLS is removed, review the changes for the first section as a whole. Excluding added files and automatically generated files, this is a ~2400 line unified diff. Anything that's added is fine (it's a feature coming from the crypto side). Anything that's replaced by something semantically equivalent or better is fine. Anything that's removed or replaced by something inferior is suspicious.

The following command lists the changes made by the first section, excluding generated files and newly added files.
```
git diff 'HEAD^{/^Merge}~1' 'HEAD^{/^Add changelog}~1' --diff-filter=DM -- . ':!library/error.c' ':!library/version_features.c' ':!programs/test/query_config.c' ':!visualc'
```

As of https://github.com/gilles-peskine-arm/mbedtls/tree/merge-crypto-unremoved-20200304-6:

* `.gitmodules`: deliberately removed.
* `CMakeLists.txt`: the only removals are references to the crypto submodule. Changes are ok.
* `Makefile`: the only removals are references to the crypto submodule. Changes are ok.
* `configs/config-no-entropy.h`: deliberate change made via crypto.
* `crypto`: deliberately removed.
* `docs/architecture/Makefile`: only additions.
* `doxygen/mbedtls.doxyfile`: only removing references to the crypto submodule.
* `include/CMakeLists.txt`:
    * :warning: https://github.com/ARMmbed/mbedtls/pull/2949 is being undone. Applying it causes every use of `cmake` to fail because at the time `target_include_directories` is executed, the `mbedtls` target is not yet declared (it's declared by `library/CMakeLists.txt`). This works in mbedtls because `add_subdirectory(include)` comes before `add_subdirectory(library)`. But in mbed-crypto the `add_subdirectory` statements are in the opposite order since 6ea2dea1c5aa1f1cc6ffe3f7ef7c56a2259a2668.
    * Additions are ok.
* `include/mbedtls/check_config.h`: deliberate changes made via crypto.
* `include/mbedtls/config.h`: deliberate changes made via crypto, removal of `crypto/` from paths, and enable more PSA options by default. The removal of a line containing `MBEDTLS_CTR_DRBG_USE_128_BIT_KEY` is because it has moved further up (deliberate change made via crypto).
* `library/CMakeLists.txt`:
    * `md_wrap.c`: removed via crypto.
    * clauses about libraries: redundant instructions have been removed.
* `library/Makefile`:
    * Removal of references to the crypto submodule. Crypto objects are now listed explicitly. To validate that the list is complete, run `make lib` and check that each file `library/*.c` acquires a corresponding `.o`.
    * The `clean` target will be fixed in a subsequent commit.
* `programs/.gitignore`: only additions.
* `programs/CMakeLists.txt`: only an addition and a cosmetic change.
* `programs/Makefile`: only removing references to the crypto submodule.
* `programs/fuzz/Makefile`: only removing references to the crypto submodule.
* `scripts/config.py`: removing `MBEDTLS_MEMORY_xxx` in `baremetal` is a deliberate change made in crypto. It's redundant with removing it in `full`.
* `scripts/data_files/vs2010-main-template.vcxproj`: deliberate change made in crypto.
* `scripts/generate_errors.pl`: the only removals are references to the crypto submodule. Changes are ok.
* `scripts/generate_visualc_files.pl`: Changes made and reviewed via crypto. Nothing extra is needed for tls.
* `tests/.gitignore`: only additions.
* `tests/CMakeLists.txt`: only additions.
* `tests/Makefile`:
    * Removal of references to the crypto submodule.
    * `APPS` is no longer filtered.
* `tests/scripts/all.sh`: was reviewed in "Restore full tls coverage to all.sh".
* `tests/scripts/check-names.sh`: deliberate changes made in crypto.
* `check-test-cases.py`: deliberate changes made in crypto, and removal of references to the crypto submodule.
* `tests/scripts/depends-hashes.pl`: deliberate changes made in crypto.
* `tests/scripts/list-enum-consts.pl`: deliberate changes made in crypto.
* `tests/scripts/list-identifiers.sh`: deliberate changes made in crypto.
* `tests/scripts/list-macros.sh`: deliberate changes made in crypto.
* `tests/scripts/run-test-suites.pl`: only removing references to the crypto submodule.
* `tests/scripts/test-ref-configs.pl`: only additions.
* `tests/suites/helpers.function`: only additions.
* `tests/suites/host_test.function`: deliberate changes made in crypto.
* `tests/suites/main_test.function`: only additions.
* `tests/suites/target_test.function`: deliberate changes made in crypto.

### Review tips

Some useful ways to refer to commits involved in the merge when working on a subsequent commit:
* The merge commit is `:/^Merge` or `HEAD^{/^Merge}`.
* Its tls precedessor is `HEAD^{/^Merge}~1`.
* Its crypto precedessor is `HEAD^{/^Merge}^2`.
* The tip of mbed-crypto as used for the merge is `$(git merge-base upstream-crypto/development 'HEAD^{/^Merge}^2')`.

We're interested in the changes to existing files. There are no removals, and all additions are to tls-only files that should stay as they are. The following command lists changes to existing files in the merge commit, excluding automatically generated files and all changed binary files:
```
git diff 'HEAD^{/^Merge}~1' 'HEAD^{/^Merge}' --diff-filter=M -- . ':!library/error.c' ':!library/version_features.c' ':!programs/test/query_config.c' ':!visualc' ':!*.pdf' ':!*.der'
```

The following command lists changes to existing files since immediately before the merge and until the current head.
```
git diff 'HEAD^{/^Merge}~1' HEAD --diff-filter=M -- . ':!library/error.c' ':!library/version_features.c' ':!programs/test/query_config.c' ':!visualc'
```

For example, to look for changes that remove X509/TLS functionality:
```
git diff 'HEAD^{/^Merge}~1' HEAD --diff-filter=M -- . ':!library/error.c' ':!library/version_features.c' ':!programs/test/query_config.c' ':!visualc' ':!*.pdf' ':!*.der' | grep -E "^-.*MBEDTLS_(ERR_)?(PKCS11|X509|NET|SSL)_"
```

The following command lists the files that existed on both sides.
```
comm -12 <(git ls-tree -r --name-only 'HEAD^{/^Merge}~1') <(git ls-tree -r --name-only $(git merge-base upstream-crypto/development 'HEAD^{/^Merge}^2'))
```

### Status

CI passes, except for Mbed OS which was broken before. The Mbed OS importer will need to be updated; this is beyond the scope of the current pull request. Known issues:

* https://github.com/ARMmbed/mbedtls/pull/2949 is undone (see `include/CMakeLists.txt` above).

All the relevant information should be in the description of this or companion PR on GitHub, or in git commit messages. But just in case here are the notes I took while working on this: 
[merge-mbed-crypto-202002.md](https://github.com/ARMmbed/mbedtls/files/4318966/merge-mbed-crypto-202002.md.txt)

I made a separate PR https://github.com/ARMmbed/mbedtls/pull/3094 to update independent documentation files such as `README.md`.

### CODE REVIEW

File: tests/scripts/basic-in-docker.sh [hunk 1]
```diff
@@ -42,6 +42,9 @@ for compiler in clang gcc; do
     run_in_docker -e CC=${compiler} make
     run_in_docker -e CC=${compiler} make test
     run_in_docker programs/test/selftest
+    run_in_docker -e OSSL_NO_DTLS=1 tests/compat.sh
```
Comment (mpg) [inline tests/scripts/basic-in-docker.sh]: Note: the end of the commit message should probably read "This only turns up changes in `basic-in-docker.sh`" rather than `run_in_docker.sh`.

**Comment by mpg** [comment]
I checked that the list of components in `all.sh` after the merge is the union of the lists of components on both side before the merge, with one exception: `component_test_full_make_gcc_o0` was removed. This is documented in the [commit message](https://github.com/ARMmbed/mbedtls/pull/3085/commits/26ff88c3eae087872670fbcafcaa10eb5d414ea7), and I double-checked that the component removed was indeed testing a strict subset of what `component_test_gcc_opt` does.

**Comment by mpg** [comment]
I also checked the **coverage of the test suites** by building in the default config and then running `scripts/run-test-suites.pl -v 2` from the `tests` (and when applicable `crypto/tests`) directory in 3 branches: this PR's branch, the merge base on the mbedtls side, and crypto/development, and compared the results.

Comparing this branch to the merge-base on the mbedtls side:

- the set of test suites is the same on both sides (tested with cmake too)
- the total number of tests available is the same on both sides
- however there are differences in tests run/skipped: some tests that were skipped in development are no longer skipped after the merge. More precisely: `test_suite_psa_crypto` +10 tests run, `test_suite_psa_crypto_persistent_key` +29 (all of them), `test_suite_psa_crypto_slot_management` +31, `test_suite_psa_its` +22 (all of them). These are explained by the fact that `MBEDTLS_PSA_CRYPTO_STORAGE_C` and `MBEDTLS_PSA_ITS_FILE_C` are now enabled by default (as they already were on the crypto side).

Comparing this branch to development on the crypto side:

- the set of test suites in this branch is a superset of that on the crypto side
- the number of tests available is the same on both sides for suites that exist on the crypto side
- however there are differences in tests run/skipped: some tests that were run on the crypto side are now skipped after the merge. More precisely: `test_suite_cmac` all tests skipped, `test_suite_psa_crypto` 10 more tests skipped, `test_suite_psa_crypto_metadata` 4 more tests skipped. All of these are explained by the fact that `MBEDTLS_CMAC_C` is disabled by default on the TLS side (for ABI/API stability reasons).

So I'm satisfied that the coverage of the test suites is preserved by the merge.

**Comment by mpg** [comment]
I've reviewed [the merge commit](https://github.com/ARMmbed/mbedtls/pull/3085/commits/1c234c85638348e15b65eb7d138b3dd4ef9655ef) by reproducing it locally an comparing the results.

1. I took the same parents, that I had previously verified as being correct choices, then ran `git merge` and didn't commit yet.

2. In order to make sure I understood it, I rewrote the shell script suggest in the commit message and ran it:

```sh
#!/bin/sh

# assumes the following branches exist:
# - tls-dev pointing to mbedtls/development as of March 5
# - crypto-dev pointing to mbed-crypto/development as of March 4

git ls-tree -r --name-only crypto-dev | sort > files-crypto-side
git ls-tree -r --name-only tls-dev | sort > files-tls-side
diff files-crypto-side files-tls-side | sed -n 's/^> //p' > files-tls-only

git checkout tls-dev -- $(cat files-tls-only)
git add -- $(cat files-tls-only)
```

3. I checked individually each of the files (`.github/issue_template.md`, `.travis.yml`, `README.md`, `doxygen/input/doc_mainpage.h`, `tests/.jenkins/Jenkinsfile`, `tests/data_files/Makefile`) documented as taking the TLS version in the commit message, agreed that taking the TLS version is the right thing to do, and did that as well.

4. I then resolved the remaining conflicts manually and check it the comments in the commit message matched the resolution I came up with.

5. I then committed the result and compared to the merge commit in this PR.

I only found 3 significant differences, where the code in the PR didn't match mine, and actually didn't match the comments in the commit message either:

1. In `library/CMakeLists.txt` the commit also kept the definition of `src_crypto` from the TLS side (no need to depend on `USE_CRYPTO_SUBMODULE`). This is correct, but not documented in the commit message's summary of conflict resolution.

2. In `programs/Makefile` the commit also keeps stuff about `fuzz` and `ssl_pthread_server`.This is correct, but not documented in the commit message's summary of conflict resolution.

3. In `tests/scripts/all.sh` the commit didn't quite keep the TLS version of `component_test_no_use_psa_crypto_full_cmake_asan`, contrary to what the commit message says. That's incorrect, but was fixed in a later commit ([Restore full tls coverage to all.sh](https://github.com/ARMmbed/mbedtls/pull/3085/commits/26ff88c3eae087872670fbcafcaa10eb5d414ea7)") so the only issue here is the discrepancy with the commit message.

In summary, I fully verified the merge commit and didn't find any issues beyond some minor discrepancies in the commit message.

**Comment by mpg** [comment]
I've reproduced commit https://github.com/ARMmbed/mbedtls/pull/3085/commits/26ff88c3eae087872670fbcafcaa10eb5d414ea7 by looking at the differences in `all.sh` between the merge commit and its TLS-side parent and resolving those differences without looking at the commit message.

I then compared my result to the one in the PR and found no significant differences. I also checked that the commit messages correctly describes the changes.

**Comment by mpg** [comment]
I've reviewed the remaining commits in the PR (from "Restore X509 and SSL functionality" to "Restore X509 and SSL functionality") either by reviewing the diff (most commits), or reproducing it ("Restore X509 and SSL functionality") or relying on the CI ("Re-generate automatically generated files") and found no issues.

Now I still have to double-check that nothing was lost in the merge, ie that everything that was wanted to revert after the merge commit has been reverted (this is partially guaranteed by the fact that our tests pass and I've verified they're completed, but to be extra sure I also want to check the diff directly).

File: include/mbedtls/check_config.h [hunk 2]
```diff
@@ -674,6 +674,23 @@
 #error "MBEDTLS_SSL_DTLS_ANTI_REPLAY  defined, but not all prerequisites"
```
[the remark this answers is not in the document]
Comment (mpg) [reply include/mbedtls/check_config.h]: I think that's a leftover comment from a merge based on a previous version of the unremove-noncrypto branch, which used to introduce that duplication. Newer versions of unremove-non-crypto fixed that, so this comment is no longer accurate indeed.

**Comment by mpg** [comment]
I finished reviewing the diff for "the first section" as defined in the PR cover, and the only issue I found was the inconsistent use of `target_include_directories()` in `library/CMakeLists.txt` already noticed by Ronald and that indeed seems related to #2949 though I haven't started investigating yet.

I also reviewed the diff of the `Makefile`s (in `.`, `library`, `tests`, `programs`) and the CMakeLists.txt (in `.`, `library`, `tests`, `programs` and `include`) against their version in 2.16 (which never had a crypto submodule) and found that all of the differences were justified by the underlying differences between development and 2.16 (addition of PSA, fuzz, allowing declarations after statements, etc.) except perhaps one: the version of `library/Makefile` in this PR has this extra line:
```
libmbedcrypto.$(DLEXT): | libmbedcrypto.a
```
which is only present for libmbedcrypto and whose purpose is not entirely clear to me. After investigating we'll probably want to either remove that line or add a similar ones for the other two libraries.

I've now done all the reviewing I wanted to do.

**Comment by mpg** [comment]
@ronald-cron-arm Thanks for your review! I think some of [the points you raised](https://github.com/ARMmbed/mbedtls/pull/3085#pullrequestreview-375842839) are actually OK, so I'll try explaining why I think so and you can tell me if that was convincing :)

---

In `include/mbedtls/config.h`:
```
-//#define MBEDTLS_CTR_DRBG_USE_128_BIT_KEY /**< Use 128-bit key for CTR_DRBG - may reduce security (see ctr_drbg.h) */
```
The define now comes earlier in the file (line 698) and the [commit message]() notes:
```
        * `MBEDTLS_CTR_DRBG_USE_128_BIT_KEY` single line: don't restore it since there is now a full paragraph in the proper section above.
```
The change comes from [this commit](https://github.com/ARMmbed/mbed-crypto/commit/1540e5bd04e463e83870f88e8531582fbfd4575d) in mbed-crypto.

---

In `tests/scripts/all.sh`:

```
- scripts/config.py unset MBEDTLS_MEMORY_BUFFER_ALLOC_C
```
Just before that we invoke `scripts/config.py full`, which used to set `MBEDTLS_MEMORY_BUFFER_ALLOC_C` but was [recently changed](https://github.com/ARMmbed/mbedtls/pull/2469) to leave it unset, so we no longer need to unset it explicitly. (Well to be precise `config.py full` never set that macro, but its predecessor `config.pl full` did.)

---

In `scripts/generate_visualc_files.pl`:
```
-# Need windows line endings!
-my $include_directories = <<EOT;
-../../include\r
-EOT
```
This is now handled differently (and didn't need any line ending after all):
```
my @include_directories = qw(
    include
    3rdparty/everest/include/
    3rdparty/everest/include/everest
    3rdparty/everest/include/everest/vs2010
    3rdparty/everest/include/everest/kremlib
);
my $include_directories = join(';', map {"../../$_"} @include_directories);
```

File: tests/scripts/all.sh [hunk 3]
```diff
@@ -24,7 +24,8 @@
 # configurations, and can and will arbitrarily change the current CMake
```
[the remark this answers is not in the document]
Comment (mpg) [reply tests/scripts/all.sh]: I think the problem is in the commit message. According to `config.h`, `MBEDTLS_PSA_CRYPTO_SE_C` depends on `MBEDTLS_PSA_CRYPTO_STORAGE_C` which is unset here, so `SE` should be unset as well.

**Comment by mpg** [comment]
Ok, so let me try to sum up the issues found so far (assuming Ronald agrees with my observations above).

**Issues in the tree itself**

- Incomplete sentence in the ChangeLog - obvious fix: complete the sentence
- Inconsistent use of `target_include_directories` in `library/CMakeLists.txt` - needs investigating, might fix it later
- Inconsistent extra target `libmbedcrypto.$(DLEXT): | libmbedcrypto.a` in `library/Makefile` - needs investigating, might fix it later
- As Gilles noted, reversal of #2949 - probably want to leave it for later, possibly wait for external contribution

**Issues in commit messages**

- s/shared memory/shared library/
- s/run_in_docker.sh/basic-in-docker.sh/
- s/prograns/programs/
- remove "`component_test_no_platform`: remove the unsetting of SE support." (as unsetting of SE isn't removed in the code)
- remove ""`MBEDTLS_PARAM_FAILED`: this has been duplicated." the duplication has been fixed earlier
- mention `src_crypto` in the message of the merge commit
- mention `fuzz` and `ssl_pthread_server` in the message of the merge commit

**Comment by mpg** [comment]
Ok, so regarding `libmbedcrypto.$(DLEXT): | libmbedcrypto.a` it was introduced in https://github.com/ARMmbed/mbedtls/pull/2884/commits/8af3923815643057b16567ed11792a575d867b7a and my understanding that we probably want to use it for the x509 and tls libraries as well. In a sense the inconsistency was pre-existing to the merge (we should have side-ported the fix from mbed-crypto to mbedtls earlier), but it probably makes sense to fix it in a subsequent commit in this PR anyway.

**Comment by ronald-cron-arm** [comment]
Hi, I’ve been through my comments and your replies and I agree with your summary concerning those.

Thanks, Ronald.

File: tests/scripts/all.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tests/scripts/all.sh]: Fixed the type and removed "* `component_test_no_platform`: remove the unsetting of SE support." from the commit message.

File: include/mbedtls/check_config.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply include/mbedtls/check_config.h]: Removed "* `MBEDTLS_PARAM_FAILED`: this has been duplicated. Do remove the version from crypto below, because the version from tls above is slightly improved." from the commit message.

**Comment by ronald-cron-arm** [comment]
### Change log of the "force-push"
The previous version of the PR can be found here: https://github.com/ronald-cron-arm/mbedtls/tree/merge-crypto-unremoved-20200304.

* Use new version of ARMmbed/mbed-crypto#384 for the merge commit

* Commit "Merge mbed-crypto into mbedtls: the merge commit"
    - Align the code with the commit message regarding "component_test_no_use_psa_crypto_full_cmake_asan" (take the tls version)
    - Add "* Keep the TLS definition of `src_crypto`" in conflicts resolution of "library/CMakeLists.txt"
    - Add "* Keep the `fuzz` target defined only in tls version." and
        "* Keep the recipe (only in tls version) cleaning `ssl_pthread_server` stuff for the `clean` target." in conflicts resolution of "programs/Makefile"

* Commit "Restore full tls coverage to all.sh"
   - Replace "prograns" by "programs"
   - Remove "`component_test_no_use_psa_crypto_full_cmake_asan` (`config.py`): do unset `MBEDTLS_PSA_CRYPTO_C`."
    - Remove "* `component_test_no_use_psa_crypto_full_cmake_asan` (`ssl-opt.sh`): do change the comments to include “full minus”."
    - Remove "* `component_test_no_platform`: remove the unsetting of SE support."

* Commit "Restore full tls coverage to remaining scripts"
    - Replace "run-in-dockers.sh" by "basic-in-dockers.sh"

* Commit "Restore X509 and SSL functionality"
    - Remove "* MBEDTLS_PARAM_FAILED: this has been duplicated. Do remove the version from crypto below, because the version from tls above is slightly improved." from the commit message.

* Commit "Fix the shared library build with CMake"
    - Replace "Fix the shared memory build with CMake" by
    - "Fix the shared library build with CMake" as commit title.

* Commit "Add changelog entries about the crypto submodule"
    - Add missing word at the end of sentence.

**Comment by mpg** [comment]
I've reviewed the force-push and I'm satisfied it does what it's supposed to do (and nothing more), with one exception: `Remove "* component_test_no_platform: remove the unsetting of SE support."` - I'm still seeing that line in the commit message. Otherwise this is looking very good.

**Comment by ronald-cron-arm** [comment]
I am removing "Remove "* component_test_no_platform: remove the unsetting of SE support." and will force-push again.

**Comment by mpg** [comment]
Note: all tests passed in the CI, except for the ones depending on mbed-os which are a known issue independent from this PR.

**Comment by gilles-peskine-arm** [comment]
I force-pushed https://github.com/gilles-peskine-arm/mbedtls/tree/merge-crypto-unremoved-20200304-10 which is identical to the previous version except for whitespace changes in the commit message of the crypto-tls merge commit.

**Comment by gilles-peskine-arm** [comment]
I pushed a merge of the target branch into the PR branch with the resolution of a trivial conflict (add/add in `ChangeLog`) so that we can do a final round of review and CI on this delicate PR.

**Comment by gilles-peskine-arm** [comment]
From a fresh clone of `mbedtls/development`:
```
$ git checkout myfork/merge-crypto-unremoved-20200304
warning: unable to rmdir crypto: Directory not empty
Previous HEAD position was 1fce45d... Merge pull request #3103 from AndrzejKurek/ssl-read-improve-ret-code-doc
HEAD is now at 3ed2de9... Refresh generated files
```

Maybe I didn't do the submodule removal correctly. Can we arrange to remove `crypto` if it doesn't contain anything other than the pristine files from the submodule?

**Comment by mpg** [comment]
I can't reproduce the problem. If I do a fresh clone of `mbedtls`, then check out `development` and run `git submodule update --init`, then check out this branch, I get no error and the crypto directory gets removed.

**Comment by mpg** [comment]
I checked the CI results and everything passed except for known issues with mbed-os and IAR.

**Comment by mpg** [comment]
I tried another way and still can't get any issue with getting the submodule removed:
```
git clone --recursive https://github.com/ARMMbed/mbedtls mbedtls-fresh
cd mbedtls-fresh
ls crypto
git fetch origin 3ed2de9f54f0538225feaadb722edc56439ccc85
git checkout -b merge-back-crypto 3ed2de9f54f0538225feaadb722edc56439ccc85
ls crypto
```
None of the git commands throws any error, the first `ls crypto` show the expected contents, and the second one errors out with `No such file or directory` as expected.

Are you sure your clone was really fresh? Could any of your tools be creating files behind your back? What files are left in your `crypto` directory after checking out this branch?

**Comment by gilles-peskine-arm** [comment]
```
{R1}/tmp % cat mpg.sh 
#!/bin/sh
set -ex
git clone --recursive https://github.com/ARMMbed/mbedtls "$1"
cd "$1"
ls crypto
git fetch origin 3ed2de9f54f0538225feaadb722edc56439ccc85
git checkout -b merge-back-crypto 3ed2de9f54f0538225feaadb722edc56439ccc85
ls crypto
{R1}/tmp % ./mpg.sh bionic
+ git clone --recursive https://github.com/ARMMbed/mbedtls bionic
Cloning into 'bionic'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 97556 (delta 0), reused 1 (delta 0), pack-reused 97553
Receiving objects: 100% (97556/97556), 43.45 MiB | 1.10 MiB/s, done.
Resolving deltas: 100% (75427/75427), done.
Submodule 'crypto' (https://github.com/ARMmbed/mbed-crypto) registered for path 'crypto'
Cloning into '/tmp/bionic/crypto'...
remote: Enumerating objects: 81932, done.        
remote: Total 81932 (delta 0), reused 0 (delta 0), pack-reused 81932        
Receiving objects: 100% (81932/81932), 23.75 MiB | 2.38 MiB/s, done.
Resolving deltas: 100% (66192/66192), done.
Submodule path 'crypto': checked out 'cf4a40ba0a3086cabb5a8227245191161fd26383'
+ cd bionic
+ ls crypto
3rdparty               LICENSE         configs  library   visualc
CMakeLists.txt         Makefile        docs     programs
CONTRIBUTING.md        README.md       doxygen  scripts
DartConfiguration.tcl  apache-2.0.txt  include  tests
+ git fetch origin 3ed2de9f54f0538225feaadb722edc56439ccc85
remote: Enumerating objects: 4355, done.
remote: Counting objects: 100% (4355/4355), done.
remote: Compressing objects: 100% (83/83), done.
remote: Total 14668 (delta 4284), reused 4329 (delta 4272), pack-reused 10313
Receiving objects: 100% (14668/14668), 5.44 MiB | 1.40 MiB/s, done.
Resolving deltas: 100% (11392/11392), completed with 455 local objects.
From https://github.com/ARMMbed/mbedtls
 * branch                3ed2de9f54f0538225feaadb722edc56439ccc85 -> FETCH_HEAD
+ git checkout -b merge-back-crypto 3ed2de9f54f0538225feaadb722edc56439ccc85
warning: unable to rmdir 'crypto': Directory not empty
Switched to a new branch 'merge-back-crypto'
+ ls crypto
3rdparty               LICENSE         configs  library   visualc
CMakeLists.txt         Makefile        docs     programs
CONTRIBUTING.md        README.md       doxygen  scripts
DartConfiguration.tcl  apache-2.0.txt  include  tests
{R1}/tmp % git --version
git version 2.17.1
```

That's under Ubuntu 18.04. I'd originally tried Ubuntu 16.04 with the same result.

**Comment by mpg** [comment]
Using git 2.25.0 (compiled and installed locally) here, in an Ubuntu 16.04 environment. There might also be differences in our git configs that matter (for example, I have `submodule.recurse=true`, which was one reason why I installed something more recent than what 16.04 provides). I'll try a few variants and see how it goes.

**Comment by mpg** [comment]
Ok, after testing I can confirm that changing `submodule.recurse` to false in my global gives me the same error as you.

I think we'll just have to tell people (including ourselves) to either use git >= 2.14 with `submodule.recurse=true`, or manually `rm -r crypto` after updating. Note that people who have an unclean tree (as a result of building, running tests, etc) are likely to need to `rm -r crypto` manually anyway.

**Comment by gilles-peskine-arm** [comment]
This has been approved in review, and CI is passing as much as it can these days.

There is a potential regression with respect to include directories when using Mbed TLS as a CMake subproject, as noted [here](https://github.com/ARMmbed/mbedtls/pull/2949#issuecomment-597175560) and tracked in https://github.com/ARMmbed/mbedtls/issues/3108. Since it isn't breaking anything covered by our testing, we aren't sure whether there is an actual regression, and this is a build-time issue which cannot go undetected and should have possible workarounds, the risk of regression is acceptable.

So I'm going ahead and merging.
